common/pango.c

```diff
@@ -152,3 +152,33 @@ void render_text(cairo_t *cairo, const char *font,
 	g_object_unref(layout);
 	free(buf);
 }
+
+void render_text_ellipsize(cairo_t *cairo, const char *font, double scale,
+		bool markup, double width, const char *fmt, ...) {
+	va_list args;
+	va_start(args, fmt);
+	// Add one since vsnprintf excludes null terminator.
+	int length = vsnprintf(NULL, 0, fmt, args) + 1;
+	va_end(args);
+
+	char *buf = malloc(length);
+	if (buf == NULL) {
+		sway_log(SWAY_ERROR, "Failed to allocate memory");
+		return;
+	}
+	va_start(args, fmt);
+	vsnprintf(buf, length, fmt, args);
+	va_end(args);
+
+	PangoLayout *layout = get_pango_layout(cairo, font, buf, scale, markup);
+	pango_layout_set_ellipsize(layout, PANGO_ELLIPSIZE_END);
+	pango_layout_set_width(layout, width * PANGO_SCALE);
```

Comment:
Since this is basically the same as `render_test` (just adding those two lines) and it in general makes sense to ellpisize when the text is too wide, why not just patch `render_text`? The width can probably queried from the cairo context, too.

Reply:
Yes, sure. Sounds like a good idea.